propagator/propagator.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagator
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// TraceContextHeaderName is the HTTP header field for Google Cloud Trace
+// https://cloud.google.com/trace/docs/setup#force-trace
+const TraceContextHeaderName = "x-cloud-trace-context"
+
+// TraceContextHeaderFormat is the regular expression pattan for valid Cloud Trace header value
+const TraceContextHeaderFormat = "(?P<trace_id>[0-9a-f]{32})/(?P<span_id>[0-9]{1,20})(;o=(?P<trace_flags>[0-9]))?"
+
+// TraceContextHeaderRe is a regular expression object of TraceContextHeaderFormat.
+var TraceContextHeaderRe = regexp.MustCompile(TraceContextHeaderFormat)
+
+var traceContextHeaders = []string{TraceContextHeaderName}
+
+// CloudTraceFormatPropagator is a propagator for Cloud Trace format.
+var CloudTraceFormatPropagator = &cloudTraceFormatPropagator{}
+
+type cloudTraceFormatPropagator struct{}
+
+func (p *cloudTraceFormatPropagator) getHeaderValue(carrier propagation.TextMapCarrier) string {
+	header := carrier.Get(TraceContextHeaderName)
+	if header != "" {
+		return header
+	}
+
+	for _, key := range carrier.Keys() {
+		if strings.ToLower(key) == TraceContextHeaderName {
+			header = carrier.Get(key)
+			if header != "" {
+				return header
+			}
+		}
+	}
+	return ""
+}
+
+// Inject injects a context to the carrier following Google Cloud Trace format.
+func (p *cloudTraceFormatPropagator) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	span := trace.SpanFromContext(ctx)
+	sc := span.SpanContext()
+
+	// https://cloud.google.com/trace/docs/setup#force-trace
+	// Trace ID: 32-char hexadecimal value representing a 128-bit number.
+	// Span ID: decimal representation of the unsigned interger.
+	sidHex := sc.SpanID().String()
+	sid, err := strconv.ParseUint(sidHex, 16, 64)
+	if err != nil {
+		return
+	}
+	header := fmt.Sprintf("%s/%d;o=%s",
+		sc.TraceID().String(),
+		sid,
+		sc.TraceFlags().String(),
+	)
+	carrier.Set(TraceContextHeaderName, header)
+}
+
+// Extract extacts a context from the carrier if the header contains Google Cloud Trace header format.
+func (p *cloudTraceFormatPropagator) Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
```

Comment:
QQ:  How does the Go SDK handle precedence rules here?

If a trace is propogated via XCTC and another via W3C,  which one wins?   In Java, we had to add EXPLICIT "If a trace is already available in Context, don't do anything" here due to lacking specification around this scenario of mulitple propagators.    Can you confirm whether the SDK handles the case of multiple propagators or whether or not you need to supress this extract if a previous propagator has already pulled context?

Reply:
Good question. As far as I confirmed the upstream contrib propagators for b3 and jaeger, they don't have such a validation, so I followed those.

* https://github.com/open-telemetry/opentelemetry-go-contrib/blob/05ef436536dc499361b6c9d3546662a99c2f918f/propagators/jaeger/jaeger_propagator.go#L84-L157
* https://github.com/open-telemetry/opentelemetry-go-contrib/blob/05ef436536dc499361b6c9d3546662a99c2f918f/propagators/b3/b3_propagator.go#L133-L161

My understanding is that the only global propagator always wins and other format will be ignored if we don't specify the format with `otel.NewCompsiteTextMapPropagator`.

Reply:
Right, but that doesn't mean all of these don't suffer from this issue (which we've seen consistently show up in all SDKs).  It looks like Go has a "last one wins" philsophy basaed on this code: https://github.com/open-telemetry/opentelemetry-go/blob/2b1bb29ef89708de6a0676d3bf7aa6162594cf35/propagation/propagation.go#L95

This means *any* documentation you have around XCTC should list XCTC *before* w3c trace context so that we ensure we're prioritizing attaching to the new standard vs. the legacy traces.   This should also help with Google's load balancers "cutting" traces by providing a new XCTC when you cross a LB/infra boundary.